syntax = "proto3";


package fmi_proto;

// https://developers.google.com/protocol-buffers/docs/reference/java-generated#package
option java_package = "";
option java_outer_classname = "UnifmuFmi";

// ----------------------- Common ----------------------

message UnifmuSerialize {}

message UnifmuDeserialize {
    bytes state = 1;
}

message EmptyReturn {}



// ----------------------- FMI3 ----------------------

enum Fmi3Status {
  FMI3_OK = 0;
  FMI3_WARNING=1;
  FMI3_DISCARD=2;
  FMI3_ERROR=3;
  FMI3_FATAL=4;
}


message Fmi3InstantiateModelExchange {
  string instance_name = 1;
  string instantiation_token = 2;
  string resource_path = 3;
  bool visible = 4;
  bool logging_on = 5;
}


message Fmi3InstantiateCoSimulation {
  string instance_name = 1;
  string instantiation_token = 2;
  string resource_path = 3;
  bool visible = 4;
  bool logging_on = 5;
  bool event_mode_used = 6;
  bool early_return_allowed = 7;
  repeated uint32 required_intermediate_variables = 8;
}

message Fmi3InstantiateScheduledExecution {
  string instance_name = 1;
  string instantiation_token = 2;
  string resource_path = 3;
  bool visible = 4;
  bool logging_on = 5;
}

message Fmi3DoStep {
  double current_communication_point = 1;
  double communication_step_size = 2;
  bool no_set_fmu_state_prior_to_current_point = 3;

}

message Fmi3SetDebugLogging {
  bool logging_on = 1;
  repeated string categories = 2;
}

message Fmi3EnterInitializationMode {
  optional double tolerance = 1;
  double start_time = 2;
  optional double stop_time = 3;
}

message Fmi3ExitInitializationMode {}

message Fmi3FreeInstance {}

message Fmi3Terminate {}

message Fmi3Reset {}

// ----------------------- FMI3 Getters ----------------------


message Fmi3GetFloat32 {
  repeated uint32 value_references = 1;
}

message Fmi3GetFloat64 {
  repeated uint32 value_references = 1;
}

message Fmi3GetInt8 {
  repeated uint32 value_references = 1;
}

message Fmi3GetUInt8 {
  repeated uint32 value_references = 1;
}

message Fmi3GetInt16 {
  repeated uint32 value_references = 1;
}

message Fmi3GetUInt16 {
  repeated uint32 value_references = 1;
}

message Fmi3GetInt32 {
  repeated uint32 value_references = 1;
}

message Fmi3GetUInt32 {
  repeated uint32 value_references = 1;
}

message Fmi3GetInt64 {
  repeated uint32 value_reference = 1;
}

message Fmi3GetUInt64 {
  repeated uint32 value_references = 1;
}

message Fmi3GetBoolean {
  repeated uint32 value_references = 1;
}

message Fmi3GetString {
  repeated uint32 value_references = 1;
}

message FmiGetBinary {
  repeated uint32 value_references = 1;
}

message Fmi3GetDirectionalDerivative {
  repeated uint32 unknowns = 1;
  repeated uint32 knowns = 2;
  repeated double seed = 3;
  repeated double sensitivity = 4;
}

message Fmi3GetAdjointDerivative {
  repeated uint32 unknowns = 1;
  repeated uint32 knowns = 2;
  repeated double seed = 3;
  repeated double sensitivity = 4;
}

message Fmi3GetOutputDerivatives {
  repeated uint32 value_references = 1;
  repeated uint32 orders = 2;
  repeated double values = 3;
}

// ----------------------- FMI3 Setters ----------------------

message Fmi3SetFloat32 {
  repeated uint32 value_reference = 1;
  repeated float values = 2;
}

message Fmi3SetFloat64 {
  repeated uint32 value_reference = 1;
  repeated double values = 2;
}

message Fmi3SetInt8 {
  repeated uint32 value_reference = 1;
  repeated int32 values = 2;
}

message Fmi3SetUInt8 {
  repeated uint32 value_reference = 1;
  repeated uint32 values = 2;
}

message Fmi3SetInt16 {
  repeated uint32 value_reference = 1;
  repeated int32 values = 2;
}

message Fmi3SetUInt16 {
  repeated uint32 value_reference = 1;
  repeated uint32 values = 2;
}

message Fmi3SetInt32 {
  repeated uint32 value_reference = 1;
  repeated int32 values = 2;
}

message Fmi3SetUInt32 {
  repeated uint32 value_reference = 1;
  repeated uint32 values = 2;
}

message Fmi3SetInt64 {
  repeated uint32 value_reference = 1;
  repeated int64 values = 2;
}

message Fmi3SetUInt64 {
  repeated uint32 value_reference = 1;
  repeated uint64 values = 2;
}

message Fmi3SetBoolean {
  repeated uint32 value_references = 1;
  repeated bool values = 2;
}

message Fmi3SetString {
  repeated uint32 value_references = 1;
  repeated string values = 2;
}

message FmiSetBinary {
  repeated uint32 value_references = 1;
  repeated bytes values = 2;
}

// ----------------------- FMI3 Return Values ----------------------

message Fmi3StatusReturn {
    Fmi3Status status = 1;
}

message Fmi3FreeInstanceReturn {}

message Fmi3GetFloat32Return {
  Fmi3Status status = 1;
  repeated float values = 2;
}

message Fmi3GetFloat64Return {
  Fmi3Status status = 1;
  repeated double values = 2;
}

message Fmi3GetInt8Return {
  Fmi3Status status = 1;
  repeated int32 values = 2;
}

message Fmi3GetUInt8Return {
  Fmi3Status status = 1;
  repeated uint32 values = 2;
}

message Fmi3GetInt16Return {
  Fmi3Status status = 1;
  repeated int32 values = 2;
}

message Fmi3GetUInt16Return {
  Fmi3Status status = 1;
  repeated uint32 values = 2;
}

message Fmi3GetInt32Return {
  Fmi3Status status = 1;
  repeated int32 values = 2;
}

message Fmi3GetUInt32Return {
  Fmi3Status status = 1;
  repeated uint32 values = 2;
}

message Fmi3GetInt64Return {
  Fmi3Status status = 1;
  repeated int64 values = 2;
}

message Fmi3GetUInt64Return {
  Fmi3Status status = 1;
  repeated uint64 values = 2;
}

message Fmi3GetBooleanReturn {
  Fmi3Status status = 1;
  repeated bool values = 2;
}

message Fmi3GetStringReturn {
  Fmi3Status status = 1;
  repeated string values = 2;
}

message FmiGetBinaryReturn {
  Fmi3Status status = 1;
  repeated bytes values = 2;
}

message Fmi3GetDirectionalDerivativeReturn {
  Fmi3Status status = 1;
  repeated double values = 2;
}

message Fmi3GetAdjointDerivativeReturn {
  Fmi3Status status = 1;
  repeated double values = 2;
}

message Fmi3GetOutputDerivativesReturn {
  Fmi3Status status = 1;
  repeated double values = 2;
}

message UnifmuFmi3SerializeReturn {
  Fmi3Status status = 1;
  bytes state = 2;
}

// ----------------------- FMI2 ----------------------

enum Fmi2Status {
  FMI2_OK = 0;
  FMI2_WARNING=1;
  FMI2_DISCARD=2;
  FMI2_ERROR=3;
  FMI2_FATAL=4;
  FMI2_PENDING=5;
}

enum FMI2Type {
  FMI2_MODEL_EXCHANGE = 0;
  FMI2_CO_SIMULATION = 1;
}

message Fmi2Instantiate {
  string instance_name = 1;
  FMI2Type fmu_type = 2;
  string fmu_guid = 3;
  string fmu_resource_location = 4;
  bool visible = 5;
  bool logging_on = 6; 
}

message Fmi2DoStep {
  double current_time = 1;
  double step_size = 2;
  bool no_step_prior = 3;
}

message Fmi2SetDebugLogging {
  repeated string categories = 1;
  bool logging_on = 2;
}

message Fmi2SetupExperiment {   
  double start_time = 1;
  optional double stop_time = 2;
  optional double tolerance = 3;
}

message Fmi2EnterInitializationMode {}

message Fmi2ExitInitializationMode {}

message Fmi2CancelStep {}

message Fmi2FreeInstance {}

message Fmi2Terminate {}

message Fmi2Reset {}

// ----------------------- FMI2 Getters ----------------------

message Fmi2GetReal {
  repeated uint32 references = 1;
}

message Fmi2GetInteger {
  repeated uint32 references = 1;
}

message Fmi2GetBoolean {
  repeated uint32 references = 1;
}

message Fmi2GetString {
  repeated uint32 references = 1;
}

message Fmi2GetDirectionalDerivatives {
  repeated uint32 references_unknown = 1;
  repeated uint32 references_known = 2;
  repeated double direction_known = 3;
}
message Fmi2GetRealOutputDerivatives {
  repeated uint32 references = 1;
  repeated int32 orders = 2;
}

// ----------------------- FMI2 Setters ----------------------

message Fmi2SetReal {
  repeated uint32 references = 1;
  repeated double values = 2;
}

message Fmi2SetInteger {
  repeated uint32 references = 1;
  repeated int32 values = 2;
}

message Fmi2SetBoolean {
  repeated uint32 references = 1;
  repeated bool values = 2;
}

message Fmi2SetString {
  repeated uint32 references = 1;
  repeated string values = 2;
}

message Fmi2SetRealInputDerivatives {
  repeated uint32 references = 1;
  repeated int32 orders = 2;
  repeated double values = 3; 
}

// ----------------------- FMI2 Return Values ----------------------

message Fmi2StatusReturn {
  Fmi2Status status = 1;
}

message Fmi2FreeInstanceReturn {}

message Fmi2GetRealReturn {
  Fmi2Status status = 1;
  repeated double values = 2;
}

message Fmi2GetIntegerReturn {
    Fmi2Status status = 1;
    repeated int32 values = 2;
}

message Fmi2GetBooleanReturn {
    Fmi2Status status = 1;
    repeated bool values = 2;
}

message Fmi2GetStringReturn {
    Fmi2Status status = 1;
    repeated string values = 2;
}

message Fmi2GetRealOutputDerivativesReturn {
  Fmi2Status status = 1;
  repeated double values = 2;
}

message Fmi2GetDirectionalDerivativesReturn {
  Fmi2Status status = 1;
  repeated double values = 2;
}

message UnifmuFmi2SerializeReturn {
  Fmi2Status status = 1;
  bytes state = 2;
}

// ----------------------- FMI Command Wrapper ----------------------

message FmiCommand {
  oneof command {

    // FMI3

    Fmi3InstantiateModelExchange Fmi3InstantiateModelExchange= 1;
    Fmi3InstantiateCoSimulation Fmi3InstantiateCoSimulation = 2;
    Fmi3InstantiateScheduledExecution Fmi3InstantiateScheduledExecution = 3;

    Fmi3DoStep Fmi3DoStep = 4;
    Fmi3SetDebugLogging Fmi3SetDebugLogging = 5;

    Fmi3EnterInitializationMode Fmi3EnterInitializationMode = 6;
    Fmi3ExitInitializationMode Fmi3ExitInitializationMode = 7;
    
    Fmi3FreeInstance Fmi3FreeInstance = 8;
    Fmi3Terminate Fmi3Terminate = 9;
    Fmi3Reset Fmi3Reset = 10;

    Fmi3GetFloat32 Fmi3GetFloat32 = 13;
    Fmi3GetFloat64 Fmi3GetFloat64 = 14;
    Fmi3GetInt8 Fmi3GetInt8 = 15;
    Fmi3GetUInt8 Fmi3GetUInt8 = 16;
    Fmi3GetInt16 Fmi3GetInt16 = 17;
    Fmi3GetUInt16 Fmi3GetUInt16 = 18;
    Fmi3GetInt32 Fmi3GetInt32 = 19;
    Fmi3GetUInt32 Fmi3GetUInt32 = 20;
    Fmi3GetInt64 Fmi3GetInt64 = 21;
    Fmi3GetUInt64 Fmi3GetUInt64 = 22;
    Fmi3GetBoolean Fmi3GetBoolean = 23;
    Fmi3GetString Fmi3GetString = 24;
    FmiGetBinary FmiGetBinary = 25;

    Fmi3GetDirectionalDerivative Fmi3GetDirectionalDerivative = 26;
    Fmi3GetAdjointDerivative Fmi3GetAdjointDerivative = 27;
    Fmi3GetOutputDerivatives Fmi3GetOutputDerivatives = 28;

    Fmi3SetFloat32 Fmi3SetFloat32 = 29;
    Fmi3SetFloat64 Fmi3SetFloat64 = 30;
    Fmi3SetInt8 Fmi3SetInt8 = 31;
    Fmi3SetUInt8 Fmi3SetUInt8 = 32;
    Fmi3SetInt16 Fmi3SetInt16 = 33;
    Fmi3SetUInt16 Fmi3SetUInt16 = 34;
    Fmi3SetInt32 Fmi3SetInt32 = 35;
    Fmi3SetUInt32 Fmi3SetUInt32 = 36;
    Fmi3SetInt64 Fmi3SetInt64 = 37;
    Fmi3SetUInt64 Fmi3SetUInt64 = 38;
    Fmi3SetBoolean Fmi3SetBoolean = 39;
    Fmi3SetString Fmi3SetString = 40;
    FmiSetBinary FmiSetBinary = 41;

    // FMI2

    Fmi2DoStep Fmi2DoStep = 42;

    Fmi2SetDebugLogging Fmi2SetDebugLogging = 43;

    Fmi2SetupExperiment Fmi2SetupExperiment = 44;
    
    Fmi2EnterInitializationMode Fmi2EnterInitializationMode = 45;
    Fmi2ExitInitializationMode Fmi2ExitInitializationMode = 46;
    Fmi2FreeInstance Fmi2FreeInstance = 47;

    Fmi2CancelStep Fmi2CancelStep = 48;
    Fmi2Terminate Fmi2Terminate = 49;
    Fmi2Reset Fmi2Reset = 50;

    Fmi2GetReal Fmi2GetReal = 51;
    Fmi2GetInteger Fmi2GetInteger = 52;
    Fmi2GetBoolean Fmi2GetBoolean = 53;
    Fmi2GetString Fmi2GetString = 54;

    Fmi2GetDirectionalDerivatives Fmi2GetDirectionalDerivatives = 55;
    Fmi2GetRealOutputDerivatives Fmi2GetRealOutputDerivatives = 56;

    Fmi2SetReal Fmi2SetReal = 57;
    Fmi2SetInteger Fmi2SetInteger = 58;
    Fmi2SetBoolean Fmi2SetBoolean = 59;
    Fmi2SetString Fmi2SetString = 60;

    Fmi2SetRealInputDerivatives Fmi2SetRealInputDerivatives = 61;

    Fmi2Instantiate Fmi2Instantiate = 62; 

    // Common

    UnifmuSerialize UnifmuSerialize = 63;
    UnifmuDeserialize UnifmuDeserialize = 64;
  }
}